px4_ros2_cpp/src/utils/geodesic.cpp

```diff
@@ -0,0 +1,185 @@
+/****************************************************************************
+ * Copyright (c) 2024 PX4 Development Team.
+ * SPDX-License-Identifier: BSD-3-Clause
+ ****************************************************************************/
+
+#include <px4_ros2/utils/geodesic.hpp>
+
+#include "map_projection_impl.hpp"
+
+namespace px4_ros2
+{
+
+MapProjection::MapProjection(Context & context)
+: _node(context.node())
+{
+  _map_projection_math = std::make_shared<MapProjectionImpl>();
+  _vehicle_local_position_sub = _node.create_subscription<px4_msgs::msg::VehicleLocalPosition>(
+    "/fmu/out/vehicle_local_position", rclcpp::QoS(10).best_effort(),
+    [this](px4_msgs::msg::VehicleLocalPosition::UniquePtr msg) {
+      const uint64_t timestamp = msg->ref_timestamp;
+      if (!isInitialized() && std::isfinite(timestamp) && timestamp != 0) {
+        _map_projection_math->initReference(
+          msg->ref_lat, msg->ref_lon, msg->ref_alt,
+          msg->ref_timestamp);
```

Comment:
The reference might change later on (on ekf resets). You can use `ref_timestamp` to keep track of that. We should also add a callback, so a user can get notified when that happens.

Reply:
Yes this condition. It might happen during normal operation, so I would not warn.
The additional callback is used so the user can update other local references. This is where it gets complicated and we'll need to think a bit about it. We can also do that later.